core/lib/src/rocket.rs

```diff
@@ -805,12 +809,36 @@ impl Rocket {
         // NB: executor must be passed manually here, see hyperium/hyper#1537
         let server = hyper::Server::builder(incoming)
             .executor(runtime.executor())
-            .serve(service);
+            .serve(service)
+            .with_graceful_shutdown(shutdown_receiver.unit_error().boxed().compat());
+
+        runtime.block_on(server)?;
 
-        // TODO.async: Use with_graceful_shutdown, and let launch() return a Result<(), Error>
-        runtime.block_on(server).expect("TODO.async handle error");
+        Ok(())
+    }
 
-        unreachable!("the call to `block_on` should block on success")
+    /// Returns a [`ShutdownHandle`], which can be used to gracefully terminate
+    /// the instance of Rocket. As you can dynamically retreive a
+    /// `ShutdownHandle` in routes, this should only be used when access is
+    /// needed externally.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # #![feature(proc_macro_hygiene, async_await)]
+    /// #
+    /// let rocket = rocket::ignite();
+    /// let handle = rocket.get_shutdown_handle();
+    ///
+    /// // Likely called from another thread.
+    /// handle.shutdown();
```

Comment:
Let's make the example appear (with some invisible `if false {` magic) to start a thread that sleeps 10 seconds and then calls `handle.shutdown()`, actually calling shutdown immediately.